File: ios/Classes/FADRewarded.swift
//
//  FADRewarded.swift
//  bilmobileadsflutter
//
//  Created by HNL_MAC on 12/28/20.
//

import Flutter
import BilMobileAds

public class FADRewarded: NSObject, FlutterPlugin {
    
    fileprivate var allAds: [Int: ADRewarded] = [:]
    fileprivate var adDelegates: [Int: FADRewardedDelegate] = [:]
    fileprivate var pluginRegistrar: FlutterPluginRegistrar?
    
    fileprivate var adUnitId: String?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FADRewarded()
        instance.pluginRegistrar = registrar
        let channel = FlutterMethodChannel(name: Utils.REWARDED, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    private func isPluginReady(id: Int, callMethod: String) -> ADRewarded? {
        guard let adObj: ADRewarded = allAds[id] else {
            if callMethod != "create" {
                print("\(Utils.LOGTAG) -> ADRewarded \(id) is nil. You need init ADRewarded first.")
                return nil
            }
            return nil
        }
        return adObj
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if (call.method == "create" && !FPBMobileAds.shared.isPluginReady()) { return }
        
        guard let args = call.arguments as? [String : Any]  else {
            result(FlutterError(code: "Missing args!", message: "Unable to convert args to [String : Any]", details: nil))
            return
        }
        
        let id = args["id"] as! Int
        let rewarded = self.isPluginReady(id: id, callMethod: call.method);
        if rewarded != nil {
            print("\(Utils.LOGTAG) -> Call - Func: \(call.method) | args: \(String(describing: call.arguments != nil ? call.arguments : "")) ");
        }
        switch call.method {
        case "setListener":
            if rewarded == nil { return }
            
            let channel = FlutterMethodChannel(name: "\(Utils.REWARDED)_\(id)", binaryMessenger: pluginRegistrar!.messenger())
            let adDelegate = FADRewardedDelegate(channel: channel)
            rewarded!.setListener(adDelegate)
            adDelegates[id] = adDelegate
            break
        case "create":
            if (rewarded == nil) {
                let adUnitId = args["adUnitId"] as! String
                let viewController: UIViewController = (UIApplication.shared.delegate?.window??.rootViewController)!;
                let rewarded = ADRewarded(viewController, placement: adUnitId)
                allAds[id] = rewarded
            }
            result(nil)
            break
        case "preLoad":
            if rewarded == nil { return }
            rewarded!.preLoad()
            break
        case "show":
            if rewarded == nil { return }
            rewarded!.show()
            break
        case "isReady":
            if rewarded == nil { return }
            result(rewarded!.isReady())
            break
        case "destroy":
            if rewarded == nil { return }
            
            rewarded?.destroy()
            allAds.removeValue(forKey: id)
            adDelegates.removeValue(forKey: id)
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

class FADRewardedDelegate: NSObject, ADRewardedDelegate {
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func rewardedDidReceiveAd() {
        channel.invokeMethod("loaded", arguments: nil)
    }
    
    func rewardedDidPresent() {
        channel.invokeMethod("opened", arguments: nil)
    }
    
    func rewardedDidDismiss() {
        channel.invokeMethod("closed", arguments: nil)
    }
    
    func rewardedUserDidEarn(rewardedItem: ADRewardedItem) {
        channel.invokeMethod("rewarded", arguments: [
            "type": rewardedItem.getType(),
            "amount": rewardedItem.getAmount()
        ])
    }
    
    func rewardedFailedToLoad(error: String) {
        channel.invokeMethod("failedToLoad", arguments: error)
    }
    
    func rewardedFailedToPresent(error: String) {
        channel.invokeMethod("failedToShow", arguments: error)
    }
    
}

File: android/src/main/java/com/bil/bilmobileads/flutter/bilmobileadsflutter/FADRewarded.java
package com.bil.bilmobileads.flutter.bilmobileadsflutter;

import android.util.Log;

import com.bil.bilmobileads.ADRewarded;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class FADRewarded implements MethodChannel.MethodCallHandler {

    public static final FADRewarded instance = new FADRewarded();

    private FlutterPlugin.FlutterPluginBinding binding;
    private MethodChannel channel;

    private Map<Integer, ADRewarded> allAds;
    private Map<Integer, MethodChannel> allAdsDelegate;

    public void register(FlutterPlugin.FlutterPluginBinding pluginBinding) {
        channel = new MethodChannel(pluginBinding.getBinaryMessenger(), Utils.REWARDED);
        channel.setMethodCallHandler(this);

        allAds = new HashMap<>();
        allAdsDelegate = new HashMap<>();

        binding = pluginBinding;
    }

    public void dispose() {
        channel.setMethodCallHandler(null);
        channel = null;
        binding = null;
        allAds.clear();
        allAds = null;
        allAdsDelegate.clear();
        allAdsDelegate = null;
    }

    public ADRewarded isPluginReady(int id, String callMethod) {
        ADRewarded adObj = allAds.get(id);
        if (adObj == null) {
            if (!callMethod.equals("create"))
                Log.d(Utils.LOGTAG, "ADRewarded " + id + " is null. You need init ADRewarded first.");
            return null;
        }
        return adObj;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        if (call.method.equals("create") && !FPBMobileAds.instance.isPluginReady()) return;

        int id = call.argument("id");
        ADRewarded rewarded = isPluginReady(id, call.method);

        if (rewarded != null)
            Log.d(Utils.LOGTAG, "Call - Func: " + call.method + (call.arguments != null ? " | args: " + call.arguments : ""));
        switch (call.method) {
            case "setListener":
                if (rewarded == null) result.error(null, null, null);

                MethodChannel adChannel = new MethodChannel(binding.getBinaryMessenger(), Utils.REWARDED + "_" + id);
                allAdsDelegate.put(id, adChannel);
                rewarded.setListener(Utils.createAdRewardedListener(adChannel));

                result.success(null);
                break;
            case "create":
                if (rewarded == null) {
                    String adUnitId = (String) call.argument("adUnitId");
                    ADRewarded adRewarded = new ADRewarded(FPBMobileAds.instance.getFlutterActivity(),
                            adUnitId);
                    allAds.put(id, adRewarded);
                }

                result.success(null);
                break;
            case "preLoad":
                if (rewarded == null) result.error(null, null, null);
                rewarded.preLoad();

                break;
            case "show":
                if (rewarded == null) result.error(null, null, null);
                rewarded.show();

                break;
            case "isReady":
                if (rewarded == null) result.error(null, null, null);

                if (rewarded.isReady()) result.success(true);
                else result.success(false);

                break;
            case "destroy":
                if (rewarded == null) result.error(null, null, null);

                rewarded.destroy();
                allAds.remove(id);
                if (allAdsDelegate.get(id) != null) allAdsDelegate.remove(id);

                break;
            default:
                Log.d(Utils.LOGTAG, "Ignoring invoke from native. This normally shouldn\'t happen.");
                result.notImplemented();
                break;
        }
    }
}